opentelemetry-sdk/src/opentelemetry/sdk/_metrics/aggregation.py

```diff
@@ -14,124 +14,185 @@
 
 from abc import ABC, abstractmethod
 from collections import OrderedDict
+from enum import IntEnum
 from logging import getLogger
 from math import inf
+from threading import Lock
+from typing import Generic, Optional, Sequence, TypeVar
 
-from opentelemetry._metrics.instrument import _Monotonic
+from opentelemetry.sdk._metrics.measurement import Measurement
+from opentelemetry.sdk._metrics.point import Gauge, Histogram, PointT, Sum
 from opentelemetry.util._time import _time_ns
 
+
+class AggregationTemporality(IntEnum):
+    AGGREGATION_TEMPORALITY_UNSPECIFIED = 0
+    AGGREGATION_TEMPORALITY_DELTA = 1
+    AGGREGATION_TEMPORALITY_CUMULATIVE = 2
+
+
+_PointVarT = TypeVar("_PointVarT", bound=PointT)
+
 _logger = getLogger(__name__)
 
 
-class Aggregation(ABC):
-    @property
-    def value(self):
-        return self._value  # pylint: disable=no-member
+class Aggregation(ABC, Generic[_PointVarT]):
+    def __init__(self, is_monotonic: bool):
+        self._is_monotonic = is_monotonic
+        self._lock = Lock()
 
     @abstractmethod
-    def aggregate(self, value):
+    def aggregate(self, measurement: Measurement) -> None:
         pass
 
     @abstractmethod
-    def make_point_and_reset(self):
-        """
-        Atomically return a point for the current value of the metric and reset the internal state.
-        """
-
+    def collect(self) -> Optional[_PointVarT]:
+        pass
 
-class SumAggregation(Aggregation):
-    """
-    This aggregation collects data for the SDK sum metric point.
-    """
 
-    def __init__(self, instrument):
+class SynchronousSumAggregation(Aggregation[Sum]):
+    def __init__(self, is_monotonic: bool):
+        super().__init__(is_monotonic)
         self._value = 0
+        self._start_time_unix_nano = _time_ns()
 
-    def aggregate(self, value):
-        self._value = self._value + value
+    def aggregate(self, measurement: Measurement) -> None:
+        with self._lock:
+            self._value = self._value + measurement.value
 
-    def make_point_and_reset(self):
-        pass
+    def collect(self) -> Optional[_PointVarT]:
+        """
+        Atomically return a point for the current value of the metric and
+        reset the aggregation value.
+        """
+        now = _time_ns()
+
+        with self._lock:
+            self._value = 0
+            self._start_time_unix_nano = now + 1
+
+        return Sum(
+            aggregation_temporality=(
+                AggregationTemporality.AGGREGATION_TEMPORALITY_DELTA
+            ),
+            is_monotonic=self._is_monotonic,
+            start_time_unix_nano=self._start_time_unix_nano,
+            time_unix_nano=now,
+            value=self._value,
```

Comment:
Shouldn't this point object be prepared before reset? Otherwise you always return `0`?

Reply:
Right, fixed, added a test :v: